### PR TITLE
Project 2: Lazy code motion

### PR DESCRIPTION
**Pull request by hackedy**

### CODE REVIEW

File: content/blog/2019-10-26-lazy-code-motion/index.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+can easily claw back any performance gains due to code motion. Rather than put
+computations as early as possible, lazy code motion moves them down to the
+a later program point that still avoids redundant computation. In fact lazy code
+motion provably places computations as late as possible ([1], Theorem 3). The
+wording of this theorem is misleading--the proof grounds out in approximate
+static analysis results.
+
+# Implementation
+I implemented lazy code motion in Ocaml. Since the existing Bril infrastructure
+is written in Python, this meant I had to implement Bril JSON parsing and design
```
Comment (sampsyo) [inline content/blog/2019-10-26-lazy-code-motion/index.md]: Well, it's written in a ridiculous mishmash of Python and TypeScript.

File: content/blog/2019-10-26-lazy-code-motion/index.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+able to ameliorate these costs.
+
+On the other end of the optimization, the computation placement algorithm is
+inefficient. Lazy code motion places computations on edges of the CFG, which
+means stitching new basic blocks into edges. While they are neccessary in
+general, on many edges an inserted basic block could be safely merged with its
+predecessor or successor block. This could have performance benefits because it
+would reduce the number of jumps. Similarly the pretty-printer for CFGs
+does not omit jumps where fall-through would work--this seems like a minor
+detail but could have performance or code size implications.
```
Comment (sampsyo) [inline content/blog/2019-10-26-lazy-code-motion/index.md]: In an ideal world, there would be an off-the-shelf "CFG simplification" pass that you could run after your pass to clean up these easily-coalescable cases.

File: content/blog/2019-10-26-lazy-code-motion/index.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
++-----------------------------------------------------------------------------------------------------------------+
+|examples/lvn_test/nonlocal   |               7|                        3|             12|                       3|
++-----------------------------------------------------------------------------------------------------------------+
+|examples/dom_test/loopcond   |             117|                       46|            165|                      46|
++-----------------------------------------------------------------------------------------------------------------+
+|examples/df_test/cond        |               9|                        1|             12|                       1|
++-----------------------------------------------------------------------------------------------------------------+
+|examples/df_test/fact        |              62|                       25|             89|                      25|
++-----------------------------------------------------------------------------------------------------------------+
+```
```
Comment (sampsyo) [inline content/blog/2019-10-26-lazy-code-motion/index.md]: FWIW, I think our Markdown processor supports [tables](https://github.github.com/gfm/#tables-extension-), so you can turn this into a legit table if you want.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Very nice and clearly explained work! :tada: I have only a handful of low-level comments within.

**Comment by hackedy** [comment]
Thanks for the feedback, I tried to address it and fix the table formatting as well.

**Review by sampsyo — APPROVED** [review]
Excellent work.

**Pull request merged**